lib/ddtrace/trace_operation.rb

```diff
@@ -0,0 +1,400 @@
+require 'ddtrace/ext/priority'
+
+require 'datadog/core/environment/identity'
+
+require 'ddtrace/event'
+require 'ddtrace/span_operation'
+require 'ddtrace/trace_segment'
+require 'ddtrace/trace_digest'
+
+module Datadog
+  # Represents the act of tracing a series of operations,
+  # by generating and collecting span measurements.
+  # When completed, it yields a trace.
+  #
+  # Supports synchronous code flow *only*. Usage across
+  # multiple threads will result in incorrect relationships.
+  # For async support, a \TraceOperation should be employed
+  # per execution context (e.g. Thread, etc.)
+  #
+  # rubocop:disable Metrics/ClassLength
+  class TraceOperation
+    DEFAULT_MAX_LENGTH = 100_000
+
+    attr_accessor \
+      :agent_sample_rate,
+      :hostname,
+      :name,
+      :origin,
+      :rate_limiter_rate,
+      :resource,
+      :rule_sample_rate,
+      :sample_rate,
+      :sampling_priority,
+      :service
+
+    attr_reader \
+      :active_span_count,
+      :active_span,
+      :id,
+      :lang,
+      :parent_span_id,
+      :process_id,
+      :runtime_id,
+      :tags
+
+    attr_writer \
+      :sampled
+
+    def initialize(
+      agent_sample_rate: nil,
+      events: nil,
+      hostname: nil,
+      id: nil,
+      lang: nil,
+      max_length: DEFAULT_MAX_LENGTH,
+      name: nil,
+      origin: nil,
+      parent_span_id: nil,
+      process_id: nil,
+      rate_limiter_rate: nil,
+      resource: nil,
+      rule_sample_rate: nil,
+      runtime_id: nil,
+      sample_rate: nil,
+      sampled: nil,
+      sampling_priority: nil,
+      service: nil
+    )
+      # Attributes
+      @events = events || Events.new
+      @id = id || Utils.next_id
+      @max_length = max_length || DEFAULT_MAX_LENGTH
+      @parent_span_id = parent_span_id || 0
+      @sampled = sampled.nil? ? false : sampled
+
+      # Tags
+      @agent_sample_rate = agent_sample_rate
+
+      # TODO: Move hostname resolution elsewhere?
+      #       Don't want trace to know about configuration.
+      hostname = Core::Environment::Socket.hostname if Datadog.configuration.report_hostname
+      @hostname = hostname && hostname.empty? ? nil : hostname
```

Comment:
Can we leave hostname as `nil` if not provided, and only backfill with `Core::Environment::Socket.hostname if Datadog.configuration.report_hostname` after the trace operation has finished (and thus moving hostname resolution elsewhere)?

Reply:
Yes, we absolutely could. I just wasn't sure what to do with it. We could:

1. Leave it as is.
2. Have the `Tracer` resolve & populate this field by passing it as an arg when it creates the `TraceOperation`.
3. Have the `TraceFormatter` resolve & populate this field downstream during the worker process.

I prefer hostname resolution happens close to the canonical source: where the trace originates, for greatest accuracy. But it doesn't matter too much whether the `Tracer` or `TraceOperation` does it. Only downside to option (2) is that if you manually create a `TraceOperation`, you don't get hostname resolution out of the box. But otherwise it seems fine.

I don't like option (3) very much. It's possible that the traces somehow get processed and sent from somewhere else, or the hostname somehow changes after the trace has been completed. Not huge deal breakers, just quirks I don't like.

Reply:
Would it make sense to cache and reuse the hostname resolution? E.g. does it really need to happen anew for every trace?

Reply:
Thought about this too. If so, we could put the caching behavior to `Core::Environment::Socket.hostname` as we do with forking as well.

Reply:
Today's talk: 
> Have the Tracer resolve & populate this field by passing it as an arg when it creates the TraceOperation.

Cache hostname as well.